precise/iaf_psc_alpha_ps.cpp

```diff
@@ -0,0 +1,832 @@
+  }
+
+  if ( Time( Time::ms( t_ref_ ) ).get_steps() < 1 )
+  {
+    throw BadProperty( "Refractory time must be at least one time step." );
+  }
+  if ( tau_m_ <= 0 || tau_syn_ex_ <= 0 || tau_syn_in_ <= 0 )
+  {
+    throw BadProperty( "All time constants must be strictly positive." );
+  }
+
+  return delta_EL;
+}
+
+void
+nest::iaf_psc_alpha_ps::State_::get( DictionaryDatum& d,
+  const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, V_m_ + p.E_L_ ); // Membrane potential
+  def< double >( d, names::I_syn_ex, I_ex_ );    // Excitatory synaptic current
+  def< double >( d, names::I_syn_in, I_in_ );    // Inhibitory synaptic current
+  def< double >( d, names::dI_syn_in, dI_in_ );  // Derivative inhib. current
+  def< double >( d, names::dI_syn_ex, dI_ex_ );  // Derivative exc. current
+  def< bool >( d, names::is_refractory, is_refractory_ );
+}
+
+void
+nest::iaf_psc_alpha_ps::State_::set( const DictionaryDatum& d,
+  const Parameters_& p,
+  double delta_EL )
+{
+  if ( updateValue< double >( d, names::V_m, V_m_ ) )
+  {
+    V_m_ -= p.E_L_;
+  }
+  else
+  {
+    V_m_ -= delta_EL;
+  }
+
+  updateValue< double >( d, names::I_syn_ex, I_ex_ );
+  updateValue< double >( d, names::I_syn_in, I_in_ );
+  updateValue< double >( d, names::dI_syn_ex, dI_ex_ );
+  updateValue< double >( d, names::dI_syn_in, dI_in_ );
+}
+
+nest::iaf_psc_alpha_ps::Buffers_::Buffers_( iaf_psc_alpha_ps& n )
+  : logger_( n )
+{
+}
+
+nest::iaf_psc_alpha_ps::Buffers_::Buffers_( const Buffers_&,
+  iaf_psc_alpha_ps& n )
+  : logger_( n )
+{
+}
+
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::iaf_psc_alpha_ps::iaf_psc_alpha_ps()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::iaf_psc_alpha_ps::iaf_psc_alpha_ps( const iaf_psc_alpha_ps& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::iaf_psc_alpha_ps::init_state_( const Node& proto )
+{
+  const iaf_psc_alpha_ps& pr = downcast< iaf_psc_alpha_ps >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::iaf_psc_alpha_ps::init_buffers_()
+{
+  B_.ex_spikes_.resize();
+  B_.ex_spikes_.clear();
+  B_.in_spikes_.resize();
+  B_.in_spikes_.clear();
+  B_.currents_.clear(); // includes resize
+  B_.logger_.reset();
+
+  Archiving_Node::clear_history();
+}
+
+void
+nest::iaf_psc_alpha_ps::calibrate()
+{
+  B_.logger_.init();
+
+  V_.h_ms_ = Time::get_resolution().get_ms();
+
+  V_.psc_norm_ex_ = 1.0 * numerics::e / P_.tau_syn_ex_;
+  V_.psc_norm_in_ = 1.0 * numerics::e / P_.tau_syn_in_;
+
+  V_.gamma_ex_ = 1 / P_.c_m_ / ( 1 / P_.tau_syn_ex_ - 1 / P_.tau_m_ );
```

Comment:
these do not seem numerically stable for `tau_syn_ex_/in_ == tau_m_`, neither do they seem to be in `iaf_psc_alpha_canon`. this should be fixed, but maybe not in this PR?

Reply:
This is checked in the propagator (line 314 and below)

Reply:
hmm, as far as I understand the propagators are selected according to whether `tau_syn_ex_ ~ tau_m_` and then computed such that division by zero is avoided. I am failing to see how the same is done for the gammas? maybe I need another coffee? ;)

Reply:
No, obviously you're right! I see it now that I had mine ;)
I wouldn't mind if this was done in another PR, yes ^^"

Reply:
ok, then I would suggest to throw an exception when the time constants are too close as a temporary solution. could you additionally please create an issue that describes what should be fixed here?